pdns/dnsdistdist/doq.cc

```diff
@@ -0,0 +1,875 @@
+  for (const auto& pair : d_tlsConfig.d_certKeyPairs) {
+    auto res = quiche_config_load_cert_chain_from_pem_file(config.get(), pair.d_cert.c_str());
+    if (res != 0) {
+      throw std::runtime_error("Error loading the server certificate: " + std::to_string(res));
+    }
+    if (pair.d_key) {
+      res = quiche_config_load_priv_key_from_pem_file(config.get(), pair.d_key->c_str());
+      if (res != 0) {
+        throw std::runtime_error("Error loading the server key: " + std::to_string(res));
+      }
+    }
+  }
+
+  {
+    const std::array<uint8_t, 4> alpn{'\x03', 'd', 'o', 'q'};
+    auto res = quiche_config_set_application_protos(config.get(),
+                                                    alpn.data(),
+                                                    alpn.size());
+    if (res != 0) {
+      throw std::runtime_error("Error setting ALPN: " + std::to_string(res));
+    }
+  }
+
+  quiche_config_set_max_idle_timeout(config.get(), 5000);
+  quiche_config_set_max_recv_udp_payload_size(config.get(), MAX_DATAGRAM_SIZE);
+  quiche_config_set_max_send_udp_payload_size(config.get(), MAX_DATAGRAM_SIZE);
+  quiche_config_set_initial_max_data(config.get(), 10000000);
+  quiche_config_set_initial_max_stream_data_bidi_local(config.get(), 1000000);
+  quiche_config_set_initial_max_stream_data_bidi_remote(config.get(), 1000000);
+  quiche_config_set_initial_max_streams_bidi(config.get(), 100);
+  quiche_config_set_cc_algorithm(config.get(), QUICHE_CC_RENO);
+  // quiche_config_log_keys(config.get());
+
+  d_server_config = std::make_shared<DOQServerConfig>(std::move(config), d_internalPipeBufferSize);
+}
+
+static std::optional<PacketBuffer> getCID()
+{
+  // FIXME remplacer par notre truc de random
+  int rng = open("/dev/urandom", O_RDONLY);
+  if (rng < 0) {
+    return std::nullopt;
+
+  }
+  PacketBuffer buffer;
+  buffer.resize(LOCAL_CONN_ID_LEN);
+  auto got = read(rng, buffer.data(), LOCAL_CONN_ID_LEN);
+  if (got < 0) {
+    return std::nullopt;
+  }
+
+  return buffer;
+}
+
+static PacketBuffer mintToken(const PacketBuffer& dcid, const ComboAddress& peer)
+{
+  // FIXME: really check whether this needs to be authenticated, via HMAC for example
+  // client recoit un datagram
+  // challenge avec token
+  // suffisement d'infos pour binder a la bonne adresse
+  // filer l'original CID fille par le client.
+  // -> ne pas garder l'etat
+  // -> inclure l'info dans le token
+  // -> voir avec libsodium ?
+  // -> token plus gros avec HMAC
+  // -> regarder ce que font les autres implementations de QUIC
+  const std::array keyword = {'q', 'u', 'i', 'c', 'h', 'e'};
+  auto addrBytes = peer.toByteString();
+  PacketBuffer token;
+  token.reserve(keyword.size() + addrBytes.size() + dcid.size());
+  token.insert(token.end(), keyword.begin(), keyword.end());
+  token.insert(token.end(), addrBytes.begin(), addrBytes.end());
+  token.insert(token.end(), dcid.begin(), dcid.end());
+  return token;
+}
```

Comment:
ça il faudra vraiment regarder avant l'alpha :) Je peux m'en occuper si tu veux.

Reply:
Apparemment, un HMAC n'est pas suffisant, il faut que le contenu soit chiffré pour ne pas dévoiler d'information qui permettrait à quelqu'un qui écoute les échanges de "suivre" un client: [8.1.3. ](https://www.rfc-editor.org/rfc/rfc9000.html#section-8.1.3)[Address Validation for Future Connections](https://www.rfc-editor.org/rfc/rfc9000.html#name-address-validation-for-futu): `A token issued with NEW_TOKEN MUST NOT include information that would allow values to be linked by an observer to the connection on which it was issued. For example, it cannot include the previous connection ID or addressing information, unless the values are encrypted. `
Du coup, chiffrement avec une clé. Une idée pas idiote est de générer la clé avec un HKDF à partir de notre connection ID (qui est aléatoire) et d'un secret, qui sera configurable pour chaque `addDOQLocal` (comme ça on peut avoir un cluster de dnsdist qui utilisent la même clé). En y réfléchissant je pense que ce serait intelligent d'accepter une liste de secrets, avec un actif (utilisé pour chiffrer les nouveaux tokens) et les autres passifs (utilisés pour déchiffrer les tokens plus vieux) de manière à pouvoir faire une rotation du secret régulière tout en continuant à accepter les anciens tokens. Mais on peut commencer simple là aussi :)

Reply:
Bon, en fait Quiche ne supporte que les token de type RETRY (quand le client initie une nouvelle connection) et pas les token que le serveur peut donner au client pour éviter un RETRY à la prochaine connection (voir https://github.com/cloudflare/quiche/issues/1112) ce qui simplifie le problème pour le moment. Faisons simple.